router/internal/jobiterator/jobiterator.go

```diff
@@ -0,0 +1,166 @@
+package jobiterator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+type IteratorOptFn func(*Iterator)
+
+// WithMaxQueries sets the maximum number of queries that can be made to the jobsDB
+// for fetching more jobs.
+func WithMaxQueries(maxQueries int) IteratorOptFn {
+	return func(ji *Iterator) {
+		ji.maxQueries = maxQueries
+	}
+}
+
+// WithDiscardedPercentageTolerance sets the discarded percentage tolerance,
+// i.e. the maximum percentage of discarded jobs that can be tolerated without further querying jobsDB.
+func WithDiscardedPercentageTolerance(discardedPercentageTolerance int) IteratorOptFn {
+	return func(ji *Iterator) {
+		ji.discardedPercentageTolerance = discardedPercentageTolerance
+	}
+}
+
+// WithLegacyOrderGroupKeyFn if enabled, sets the order group key function to assume a legacy jobs pickup algorithm.
+func WithLegacyOrderGroupKey(legacy bool) IteratorOptFn {
+	return func(ji *Iterator) {
+		if legacy {
+			ji.orderGroupKeyFn = func(job *jobsdb.JobT) string { return job.LastJobStatus.JobState }
+		}
+	}
+}
+
+// Iterator is a job iterator with support for fetching more than the original set of jobs requested,
+// if case some of these jobs are discarded, according to the configured discarded percentage tolerance.
+type Iterator struct {
+	params                       jobsdb.GetQueryParamsT
+	maxQueries                   int
+	discardedPercentageTolerance int
+	orderGroupKeyFn              func(*jobsdb.JobT) string
+	getJobsFn                    func(context.Context, map[string]int, jobsdb.GetQueryParamsT, jobsdb.MoreToken) (*jobsdb.GetAllJobsResult, error)
+	state                        struct {
+		// running iterator state
+		jobs        []*jobsdb.JobT
+		idx         int
+		previousJob map[string]*jobsdb.JobT
+
+		// closed indicates whether the iterator has reached the end or not
+		closed bool
+
+		// queryCount is the number of queries made to the jobsDB
+		queryCount int
+
+		// for the next query
+		discarded         int
+		pickupMap         map[string]int
+		continuationToken jobsdb.MoreToken
+	}
+}
+
+// New returns a new job iterator
+//   - pickupMap: a map of workspaceID to number of jobs to be fetched from that workspace
+//   - params: jobsDB query parameters
+//   - getJobsFn: the function to fetch jobs from jobsDB
+//   - opts: optional iterator options
+func New(pickupMap map[string]int, params jobsdb.GetQueryParamsT, getJobsFn func(context.Context, map[string]int, jobsdb.GetQueryParamsT, jobsdb.MoreToken) (*jobsdb.GetAllJobsResult, error), opts ...IteratorOptFn) *Iterator {
+	ji := &Iterator{
+		params:                       params,
+		maxQueries:                   100,
```

Comment:
Any reason why this default is more than, what router sets by default?

Reply:
Just picked a value high enough not to be limiting by default.